go/graphd/connect.go

```diff
@@ -0,0 +1,192 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The connect portion of the graphd package facilitates connectivty to a graphd database.
+
+package graphd
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"time"
+)
+
+// Connections are sent/received on res.  awaitingConn is used as a mutex.
+type connChan struct {
+	res          chan net.Conn
+	awaitingConn chan int
+}
+
+// Returns a ready to use connCh struct.
+func initConnCh() connChan {
+	connCh := connChan{}
+	connCh.res = make(chan net.Conn)
+	connCh.awaitingConn = make(chan int)
+	return connCh
+}
+
+// dial attempts to acquire a connection to url with the specified timeout.  On connection
+// success or failure, dial will attempt to send the net.Conn over the result channel if it
+// can read from the awaitingConn channel (used as mutex here).  If the read fails (awaitingConn
+// has been closed by the caller), dial will attempt to close the acquired connection.
+func (g *graphd) dial(url *url.URL, timeout time.Duration, connCh connChan) {
+	// A timeout of zero value is equivalent to no timeout.
+	dialer := net.Dialer{Timeout: timeout}
+
+	// Attempt to dial.  Log error, but send conn anyway.  A nil conn will be discarded by the
+	// caller.
+	g.LogDebugf("dialing %v with timeout %v", url, timeout)
+	conn, err := dialer.Dial(url.Scheme, url.Host)
+	if err != nil {
+		g.LogErrf("failed to dial %v: %v", url, err)
+	}
+
+	// If an acquired connection has not yet been sent by a dilaer, go ahead and send the conn,
+	// regardless if it's nil, and return.
+	if _, ok := <-connCh.awaitingConn; ok {
+		if conn != nil {
+			g.LogDebugf("sending acquired connection to %v on channel", url)
+		} else {
+			g.LogDebugf("sending nil connection to %v on channel", url)
+		}
+		connCh.res <- conn
+		return
+	}
+
+	// A connection was already acquired by another dialer.  If we did acquire a connection,
+	// close it.
+	if conn != nil {
+		g.LogDebugf("discarding and closing acquired connection to %v", url)
+		if err := conn.Close(); err != nil {
+			g.LogErrf("failed to close acquired connection to %v, resource leak", url)
+		}
+	}
+}
+
+// isConnected checks if this graphd instance is connected to a graphd database.
+// TODO: Currently only a stub.  True case needs a write check (ie. "status()").
+func (g *graphd) isConnected() bool {
+	if g.conn == nil {
+		return false
+	}
+
+	return true
+}
+
+// Connect to a graphd database.  Connect will attempt to connect to all URLs found in the URLs
+// list associated with this graphd instance, retaining the first successful connection.  On failure,
+// an appropriate error code is returned.  If an acquired connection is already present and valid,
+// Connect returns nil.  Timeout is specified in seconds.  A timeout of 0 is treated as no timeout.
+func (g *graphd) Connect(t int) error {
```

Comment:
Can this be "Dial" ? ;) It's Go-lingua, good to stick with the terminology i think

Reply:
Sure thing, will update.